pinot-broker/src/main/java/org/apache/pinot/broker/queryquota/MaxHitRateTracker.java

```diff
@@ -0,0 +1,80 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.broker.queryquota;
+
+import com.google.common.annotations.VisibleForTesting;
+
+/**
+ * A stateful version of hit counter. Similar to the default hit counter, it maintains a list of buckets.
+ * Whereas it maintains an extra variable called _lastAccessTimestamp which tracks the last access time.
+ * If the stateful hit counter gets queried, it firstly compares the current timestamp and the last access timestamp,
+ * calculating the start index and end index among the buckets. Then, it traverses through all the valid candidate buckets.
+ * If the current timestamp has exceeded the current time range of all the buckets, this hit counter will use
+ * the current timestamp minus the default time queried time range to calculate the start time index.
+ */
+public class MaxHitRateTracker extends HitCounter {
+  private static int ONE_SECOND_BUCKET_WIDTH_MS = 1000;
+  private static int MAX_TIME_RANGE_FACTOR = 2;
+
+  private final long _maxTimeRangeMs;
+  private final long _defaultTimeRangeMs;
+  private volatile long _lastAccessTimestamp;
+
+  public MaxHitRateTracker(int timeRangeInSeconds) {
+    this(timeRangeInSeconds, timeRangeInSeconds * MAX_TIME_RANGE_FACTOR);
+  }
+
+  public MaxHitRateTracker(int defaultTimeRangeInSeconds, int maxTimeRangeInSeconds) {
```

Comment:
why do we need this constructor?

Reply:
Basically it's the `maxTimeRangeInSeconds` that needs to be passed into the parent constructor instead of the default one. Thus, we'd have to re-calculate the `maxTimeRangeInSeconds ` multiple times before we assign it to this class. Creating an extra constructor gives us the ability to reduce the duplicate calculation.

Reply:
Can we make that constructor private then?